src/UglyToad.PdfPig/Writer/TokenWriter.cs

```diff
@@ -407,40 +408,70 @@ private static void WriteStream(StreamToken streamToken, Stream outputStream)
-                for (var i = 0; i < stringToken.Data.Length; i++)
-                {
-                    var c = stringToken.Data[i];
-
-                    if (c == (char) StringStart || c == (char)StringEnd || c == (char) Backslash)
-                    {
-                        stringToken = new StringToken(stringToken.Data.Insert(i++, "\\"), stringToken.EncodedWith);
-                    }
-
-                    // Close enough.
-                    if (c > 250)
-                    {
-                        isUtf16 = true;
-                        break;
-                    }
-                }
-
-                if (isUtf16)
-                {
-                    stringToken = new StringToken(stringToken.Data, StringToken.Encoding.Utf16BE);
-                }
-            }
-
-            outputStream.WriteByte(StringStart);
-            var bytes = stringToken.GetBytes();
-            outputStream.Write(bytes, 0, bytes.Length);
-            outputStream.WriteByte(StringEnd);
-
+            {
+                // iso 88591 (or really PdfDocEncoding in non-contentstream circumstances shouldn't
+                // have these chars but seems like internally this isn't obeyed (see:
+                // CanCreateDocumentInformationDictionaryWithNonAsciiCharacters test) and it may
+                // happen during parsing as well -> switch to unicode
+                if (stringToken.Data.Any(x => x > 255))
+                {
+                    var data = new StringToken(stringToken.Data, StringToken.Encoding.Utf16BE).GetBytes();
+                    outputStream.Write(data, 0, data.Length);
+                }
+                else
+                {
+                    int ei;
+                    for (var i = 0; i < stringToken.Data.Length; i++)
+                    {
+                        var c = (int)stringToken.Data[i];
+                        if (c == (int)'(' || c == (int)')') // wastes a little space if escaping not needed but better than forward searching
+                        {
+                            outputStream.WriteByte((byte)'\\');
+                            outputStream.WriteByte((byte)c);
+                        }
+                        else if ((ei = Array.IndexOf(EscapeNeeded, c)) > -1)
+                        {
+                            outputStream.WriteByte((byte)'\\');
+                            outputStream.WriteByte((byte)Escaped[ei]);
+                        }
+                        else if (c < 32 || c > 126) // non printable
+                        {
+                            var b3 = c / 64;
+                            var b2 = (c - b3 * 64) / 8;
+                            var b1 = c % 8;
+                            outputStream.WriteByte((byte)'\\');
+                            outputStream.WriteByte((byte)(b3 + '0'));
+                            outputStream.WriteByte((byte)(b2 + '0'));
+                            outputStream.WriteByte((byte)(b1 + '0'));

```

Comment:
This could be a good use cas for bit manipulation, which should have better performances
```c#
//in this scenario, we can use | in place of +, as ZeroChar equals 0b_0011_0000, and the values are guaranteed lower than 0b_0000_1111 (line 437 + modulo 8)
outputStream.WriteByte(EscapeChar);
outputStream.WriteByte((byte)((c >> 6) | ZeroChar)); // = (c / 64) + 48
outputStream.WriteByte((byte)(((c >> 3) & 0b_0000_0111) | ZeroChar)); // = (c / 8) % 8 + 48
outputStream.WriteByte((byte)((c & 0b_0000_0111) | ZeroChar)); // = (c % 8) + 48
```

Reply:
Hadn't thought of that but if non-printable characters are serialized a lot may be worth updating with your bit logic. Quick test using BenchmarkDotNet shows improvement from ~8ns to ~5ns per char using random values in the pertinent range (0-31,128-255).